src/v/archival/ntp_archiver_service.cc

```diff
@@ -863,6 +871,24 @@ ss::future<cloud_storage::upload_result> ntp_archiver::upload_manifest(
     co_return result;
 }
 
+int64_t ntp_archiver::records_pending_upload() const {
+    const auto hwm = _parent.high_watermark();
+
+    if (hwm == model::offset{}) {
+        return 0;
+    }
+
+    const auto ot_state = _parent.get_offset_translator_state();
+    const auto kafka_hwm = ot_state->from_log_offset(hwm);
```

Comment:
This walks a btree_map, prom scrapes should generally be simple reads to avoid unwelcome scale issues.

Perhaps calculate the statistic when ntp_archer does its segment scanning loop, and accept the metric being a little out of date?

Reply:
I've ended up caching this value in the probe and only refreshing it every 30 seconds.